app/src/main/java/dracula/vlad/androcalc/MainActivity.java

```diff
@@ -113,106 +100,121 @@ public void afterTextChanged(Editable c) {
 
-        getChildViews(findViewById(R.id.ll_nums));
+        getChildViews(findViewById(R.id.numbers_linear_layout));
 
-        for (final Button btn: buttonArrayList){
-            btn.setOnClickListener(v -> {
-                String curBtnText = btn.getText().toString();
-                addAtCursorPos(curBtnText);
+        for (final Button button : buttons) {
+            button.setOnClickListener(v -> {
+                String currentButtonText = button.getText().toString();
+                addAtCursorPos(currentButtonText);
             });
         }
 
-        Button btn_Clear = findViewById(R.id.btn_Clear);
-        btn_Clear.setOnClickListener(v -> {
-            editText.getText().clear();
-            tv_Res.setText("");
+        clearAllButton = findViewById(R.id.clear_all_button);
+        clearAllButton.setOnClickListener(v -> {
+            inputView.getText().clear();
+            resultView.setText("");
         });
 
-        Button btn_Del = findViewById(R.id.btn_Del);
-        btn_Del.setOnClickListener(v -> {
-            int length = editText.getText().length();
+        deleteSingleCharButton = findViewById(R.id.delete_button);
+        deleteSingleCharButton.setOnClickListener(v -> {
+            int length = inputView.getText().length();
+
             if (length > 0) {
                 delAtCursorPos();
             }
-            if (editText.getText().length()==0)
-                tv_Res.setText("");
-        });
 
-        Button btn_Dot = findViewById(R.id.btn_Dot);
-        btn_Dot.setOnClickListener(v -> {
-            int length = editText.getText().length();
-            if (length>0){
-                int index = editText.getSelectionStart();
-                if (index!=0){
-                    String lastChar = editText.getText().toString();
-                    lastChar=String.valueOf(lastChar.charAt(index-1));
-                    Log.d(TAG, "lastChar: "+lastChar);
-                    if (!(lastChar.equals(".")))
-                        addAtCursorPos(".");
-                }
+            if (inputView.getText().length() == 0) {
+                resultView.setText("");
             }
         });
 
-        buttonArrayList.clear();
-        getChildViews(findViewById(R.id.ll_funcs));
-
-        for (final Button btn: buttonArrayList){
-
-            btn.setOnClickListener(v -> {
-
-                int lengthET = editText.getText().length();
-                String curBtn=btn.getText().toString();
+        dotInputButton = findViewById(R.id.dot_input_button);
+        dotInputButton.setOnClickListener(v -> {
+            int length = inputView.getText().length();
 
+            if (length > 0) {
+                int index = inputView.getSelectionStart();
 
-                if (!btn.getText().toString().equals("=") && lengthET!=0) {
+                if (index != 0) {
+                    String lastChar = inputView.getText().toString();
```

Comment:
inputView.getText() - multiple calls, try to rework it and avoid call the same code multiple times (see line 141)